src/lib/CollisionAvoidance/CollisionAvoidance.cpp

```diff
@@ -0,0 +1,175 @@
+				float vel_lim_y = (max_detection_distance - distance) / (max_detection_distance - MPC_COL_AVOID_D.get()) * sin(angle);
+
+				if (vel_lim_x > 0 && vel_lim_x > _move_constraints_x_normalized(1)) { _move_constraints_x_normalized(1) = vel_lim_x; }
+
+				if (vel_lim_y > 0 && vel_lim_y > _move_constraints_y_normalized(1)) { _move_constraints_y_normalized(1) = vel_lim_y; }
+
+				if (vel_lim_x < 0 && -vel_lim_x > _move_constraints_x_normalized(0)) { _move_constraints_x_normalized(0) = -vel_lim_x; }
+
+				if (vel_lim_y < 0 && -vel_lim_y > _move_constraints_y_normalized(0)) { _move_constraints_y_normalized(0) = -vel_lim_y; }
+			}
+		}
+
+	} else if (_last_message + MESSAGE_THROTTLE_US < hrt_absolute_time()) {
+		mavlink_log_critical(&_mavlink_log_pub, "No range data received");
+		_last_message = hrt_absolute_time();
+	}
+}
+
+void CollisionAvoidance::modifySetpoint(Vector2f &original_setpoint, const float max_speed)
+{
+
+	reset_constraints();
+
+	//calculate movement constraints based on range data
+	update_range_constraints();
+	_move_constraints_x = _move_constraints_x_normalized;
+	_move_constraints_y = _move_constraints_y_normalized;
+
+	// calculate the maximum velocity along x,y axis when moving in the demanded direction
+	float vel_mag = sqrt(original_setpoint(0) * original_setpoint(0) + original_setpoint(1) * original_setpoint(1));
+	float v_max_x, v_max_y;
+
+	if (vel_mag > 0.0f) {
+		v_max_x = abs(max_speed / vel_mag * original_setpoint(0));
+		v_max_y = abs(max_speed / vel_mag * original_setpoint(1));
+
+	} else {
+		v_max_x = 0.0f;
+		v_max_y = 0.0f;
+	}
+
+	//scale the velocity reductions with the maximum possible velocity along the respective axis
+	_move_constraints_x(0) *= v_max_x;
+	_move_constraints_x(1) *= v_max_x;
+	_move_constraints_y(0) *= v_max_y;
+	_move_constraints_y(1) *= v_max_y;
+
+	//apply the velocity reductions to form velocity limits
+	_move_constraints_x(0) = v_max_x - _move_constraints_x(0);
+	_move_constraints_x(1) = v_max_x - _move_constraints_x(1);
+	_move_constraints_y(0) = v_max_y - _move_constraints_y(0);
+	_move_constraints_y(1) = v_max_y - _move_constraints_y(1);
+
+	//constrain the velocity setpoint to respect the velocity limits
+	Vector2f new_setpoint = original_setpoint;
+	new_setpoint(0) = math::constrain(original_setpoint(0), -_move_constraints_x(0), _move_constraints_x(1));
+	new_setpoint(1) = math::constrain(original_setpoint(1), -_move_constraints_y(0), _move_constraints_y(1));
+
+	//warn user if collision avoidance starts to interfere
+	bool currently_interfering = (new_setpoint(0) < 0.95f * original_setpoint(0)
+				      || new_setpoint(0) > 1.05f * original_setpoint(0) || new_setpoint(1) < 0.95f * original_setpoint(1)
+				      || new_setpoint(1) > 1.05f * original_setpoint(1));
+
+	if (currently_interfering && (currently_interfering != _interfering)) {
+		mavlink_log_critical(&_mavlink_log_pub, "Collision Avoidance starts interfering");
```

Comment:
This notification is confusing and doesn't really tell the user much. OK for developers, but a "standard" user has no idea what "Collision Avoidance starts interfering" means.

@dagar Do you know what DJI does?

Reply:
@baumanta What @mhkabir is saying is that some notification is needed, but that this particular text could be improved. 

@mhkabir @dagar Based purely on [this post](https://forum.dji.com/forum.php?mod=viewthread&tid=113729) I think that DJI do a beeps with pitch/frequency based on proximity. To me that seems very familiar from car reversing sensor systems - and it would be really good if we could have similar - both QGC and on transmitter (ie via FrSky telemetry). Is this possible?
I'd also quite like a visual indicator.

Complementary text also fine. For this, what if we just used text: "Collision warning"?

A specific message for this notification probably needed?

Reply:
We plan on improving the notifications, e.g. with a visualization in QGC which shows blocked sector as it is done in cars. @mrivi uses that already for yuneec. But we do not have the capactity to do that right now, so we will stick to the message in the mean time. If you have any suggestions for improving the message, I'll be glad to change it of course.